db/db_impl_write.cc

```diff
@@ -392,15 +415,31 @@ Status DBImpl::WriteToWAL(const autovector<WriteThread::Writer*>& write_group,
     }
   }
 
+  if (concurrent) {
+    // We need to lock log_write_mutex_ since logs_ and alive_log_files might be
+    // pushed back concurrently
+    log_write_mutex_.Lock();
+  }
+  *last_sequence = versions_->FetchAddLastToBeWrittenSequence(total_count);
+  auto sequence = *last_sequence + 1;
   WriteBatchInternal::SetSequence(merged_batch, sequence);
 
   Slice log_entry = WriteBatchInternal::Contents(merged_batch);
+  log::Writer* log_writer = logs_.back().writer;
   status = log_writer->AddRecord(log_entry);
```

Comment:
Maybe whether writing to buffer should be passed in per record basis. In this case, in case, someone issues a write without going through this prepare API, it will call AddRecord() with the flush option so it should still work.